services/navigation/builtin/builtin.go

```diff
@@ -348,16 +373,32 @@ func (svc *builtIn) nextWaypoint(ctx context.Context) (navigation.Waypoint, erro
 }
 
 func (svc *builtIn) waypointReached(ctx context.Context) error {
-	wp, err := svc.nextWaypoint(ctx)
-	if err != nil {
-		return fmt.Errorf("can't mark waypoint reached: %w", err)
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	if svc.waypointInProgress == nil {
+		return errors.New("can't mark waypoint reached since there is none in progress")
 	}
-	return svc.store.WaypointVisited(ctx, wp.ID)
+	return svc.store.WaypointVisited(ctx, svc.waypointInProgress.ID)
 }
 
-func (svc *builtIn) Close(ctx context.Context) error {
-	svc.cancelFunc()
+func (svc *builtIn) callCancelFunc() {
+	svc.mu.RLock()
+	if svc.cancelFunc != nil {
+		svc.cancelFunc()
+	}
+	svc.mu.RUnlock()
 	svc.activeBackgroundWorkers.Wait()
+}
+
+func (svc *builtIn) Close(ctx context.Context) error {
+	svc.callCancelFunc()
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
```

Comment:
As with other cases, I'd suggest removing callCancelFunc() and just directly call svc.cancelFunc() (and wait) INSIDE the lock to avoid races.